proposals/ideas/xxx-flashflow.txt

```diff
@@ -0,0 +1,713 @@
+MSM_ECHO cells are the measurement traffic; the measurer generates them, sends
+them to the target, and the target echos them back. The target send a MSM_BG cells
+once per second to report the amount of background traffic it is handling. MSM_ERR
+cells are used to signal to the other party that there has been some sort of
+problem and that the measurement should be aborted. These cell types are
+described in more detail in the next section.
+
+The only cell type that is sometimes encrypted is MSM_ECHO; all other types
+are never encrypted.
+
+The relay "decrypts" MSM_ECHO cells before sending them back to the measurer; this
+mirrors the way relays decrypt/encrypt RELAY_DATA cells in order to induce
+realistic cryptographic CPU load. The measurer usually skips encrypting MSM_ECHO
+cells to reduce its own CPU load; however, to verify the relay is actually
+correctly decrypting all cells, the measurer will choose random outgoing
+cells, encrypt them, remember the ciphertext, and verify the corresponding
+incoming cell matches.
+
+3.1.3 Pre-Measurement Handshaking/Starting a Measurement
+
+The coordinator connects to the target relay and sends it a MSM_PARAMS cell.
+If the target is unwilling to be measured at this time or if the coordinator
+didn't use a TLS certificate that the target trusts, it responds with an error
+cell and closes the connection.  Otherwise it checks that the parameters of
+the measurement are acceptable (e.g. the version is acceptable, the duration
+isn't too long, etc.).  If the target is happy, it sends a MSM_PARAMS_OK,
+otherwise it sends a MSM_ERR and closes the connection.
+
+Upon learning the IP addresses of the measurers from the coordinator in the
+MSM_PARAMS cell, the target whitelists their IPs in its DoS detection
+subsystem until the measurement ends (successfully or otherwise).
+
+Upon receiving a MSM_PARAMS_OK from the target, the coordinator will instruct
+the measurers to open their TCP connections with the target. If the
+coordinator or any measurer receives a MSM_ERR, it reports the error to the
+coordinator and considers the measurement a failure. It is also a failure if
+any measurer is unable to open at least half of its TCP connections with the
+target.
+
+The payload of MSM_PARAMS cells [XXX more may need to be added]:
+
+  - version       [1 byte]
+  - msm_duration  [1 byte]
+  - num_measurers [1 byte]
+  - measurer_info [num_measurers times]
+    - ipv4_addr   [4 bytes]
+    - num_conns   [2 bytes]
+
+version dictates how this MSM_PARAMS cell shall be parsed.  msm_duration is
+the duration, in seconds, that the actual measurement will last.
+num_measurers is how many measurer_info structs follow.  For each measurer,
+the ipv4_addr it will use when connecting to the target is provided, as is
+num_conns, the number of TCP connections that measurer will open with the
+target.  Future versions of FlashFlow and this cell type will use TLS
+certificates instead of IP addresses.
```

Comment:
I am again unclear if TLS is used with the measurer connections. If not, how do you plan to use TLS certificates?

Reply:
Yes TLS is used on measurer connections like any other OR connection. The references to "no encryption" refer to no *cell* encryption. I will clarify this.